src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1996,55 +2003,69 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+
+                return BadExpression(receiverSyntax, receiver);
             }
 
-            Symbol deconstructMember = candidates[0];
+            var analyzedArguments = AnalyzedArguments.GetInstance();
+            var outVars = ArrayBuilder<OutDeconstructVarPendingInference>.GetInstance(numCheckedVariables);
 
-            // check that the deconstruct fits
-            if (deconstructMember.Kind != SymbolKind.Method)
+            try
             {
-                Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, boundRHS.Syntax, boundRHS.Type);
-                return null;
-            }
+                for (int i = 0; i < numCheckedVariables; i++)
+                {
+                    var variable = new OutDeconstructVarPendingInference(assignmentSyntax);
+                    analyzedArguments.Arguments.Add(variable);
+                    analyzedArguments.RefKinds.Add(RefKind.Out);
+                    outVars.Add(variable);
+                }
 
-            MethodSymbol deconstructMethod = (MethodSymbol)deconstructMember;
-            if (deconstructMethod.MethodKind != MethodKind.Ordinary)
-            {
-                Error(diagnostics, ErrorCode.ERR_MissingDeconstruct, boundRHS.Syntax, boundRHS.Type);
-                return null;
-            }
+                string methodName = "Deconstruct";
+                var boundExpression = BindInstanceMemberAccess(
+                                        receiverSyntax, receiverSyntax, receiver, methodName, rightArity: 0,
+                                        typeArgumentsSyntax: default(SeparatedSyntaxList<TypeSyntax>), typeArguments: default(ImmutableArray<TypeSymbol>),
+                                        invoked: true, diagnostics: diagnostics);
 
-            if (deconstructMethod.ParameterCount != numCheckedVariables)
-            {
-                Error(diagnostics, ErrorCode.ERR_DeconstructWrongParams, boundRHS.Syntax, deconstructMethod, numCheckedVariables);
-                return null;
-            }
+                boundExpression = CheckValue(boundExpression, BindValueKind.RValueOrMethodGroup, diagnostics);
+                boundExpression.WasCompilerGenerated = true;
+
+                BoundExpression result = BindInvocationExpression(
+                                            receiverSyntax, receiverSyntax, methodName, boundExpression, analyzedArguments, diagnostics, queryClause: null,
+                                            allowUnexpandedForm: true);
 
-            if (deconstructMethod.Parameters.Any(p => p.RefKind != RefKind.Out))
+                result.WasCompilerGenerated = true;
+
+                if (result.HasAnyErrors)
+                {
+                    outPlaceholders = default(ImmutableArray<BoundDeconstructValuePlaceholder>);
+                    return BadExpression(assignmentSyntax, result);
+                }
+
+                outPlaceholders = outVars.SelectAsArray(v => v.Placeholder);
```

Comment:
Yes, `AnalyzedArguments` were previously updated after overload resolution. Look at the `CoerceArguments` method, which is modified below to call `SetInferredType`. A few lines above, pre-existing code was also updating the arguments:

```C#
                if (!kind.IsIdentity)
                {
                    TypeSymbol type = GetCorrespondingParameterType(ref result, parameters, arg);
                    // [removed some code]
                    arguments[arg] = CreateConversion(argument.Syntax, argument, kind, false, type, diagnostics);
                }
                else if (argument.Kind == BoundKind.OutDeconstructVarPendingInference)
                {
                    TypeSymbol parameterType = GetCorrespondingParameterType(ref result, parameters, arg);
                    arguments[arg] = ((OutDeconstructVarPendingInference)argument).SetInferredType(parameterType, success: true);
                }
```

As for the `outVars` local, it does not get passed into any method and it is not modified. The reason I need it is because it is a more reliable way of collecting the placeholders than trying to get them from modified `AnalyzedArguments` (which is tricky in extension method scenarios).